### DISCUSSION TITLE
Add duende-identity-server7 provider

### DISCUSSION THREAD

**Discussion by MonstraG** [nextauthjs/next-auth#10036] [Ideas]
### Goals

[Recently](https://blog.duendesoftware.com/posts/20240125_v7_release/), a new version of Duende IdentityServer has been released, one of the features of which is [Pushed Authorization Requests](https://datatracker.ietf.org/doc/html/rfc9126).

### Non-Goals

n/a?

### Background

is6 is supported, so should is7

### Proposal

I'm not sure whether how next/auth operates conflicts or not with PAR, that's  the biggest question.

~~I am a bit interested in contributing.~~

**Comment by akosb** [comment]
Hi, 

 not sure this helps, but  due to some project [requirements](https://www.nhn.no/tjenester/helseid/sikkerhetskonsepter-i-kontekst-av-helseid/par-for-tjenesteeiere-og-arkitekter) I'm also looking into this.

According to the [RFC](https://www.rfc-editor.org/rfc/rfc9126.html) a PAR request will return a `request_uri` that has to be passed in to an autorizationUrl endpoint.

**NextAuth.js v4**
This version is using openid-client v5.x ([ref](https://github.com/nextauthjs/next-auth/blob/v4/packages/next-auth/package.json))
```
	"openid-client": "^5.4.0",
```
which implements a method called `pushedAuthorizationRequest` ([ref](https://github.com/panva/openid-client/blob/v5.x/docs/README.md#clientpushedauthorizationrequestparameters-extras))
Unfortunatelly there is no example provided about this methods usage, but I wrote a PoC for this:
```
	const parResponse = await client.pushedAuthorizationRequest({
		scope: "xxx xxx xxx",
		response_type: "code",
		response_mode: "form_post",
		code_challenge,
		code_challenge_method: ['S256'],		
	}, { DPoP: privateKey });

	const authUrl = await client.authorizationUrl({
		request_uri: parResponse.request_uri
	})
	return authUrl	

```
where privateKey is your JWT key and the `parResponse` looks like this:
```
parResponse: {
  request_uri: 'urn:ietf:params:oauth:request_uri:xxxxxxxxxxxxxxxxxxxxxx',
  expires_in: 600
}
```
this has to be passed in to the the `authUrl`
```
	const authUrl = await client.authorizationUrl({
		request_uri: parResponse.request_uri
	})
	return authUrl
```

Next auth does this [here](https://github.com/nextauthjs/next-auth/blob/v4/packages/next-auth/src/core/lib/oauth/authorization-url.ts) so probably this is where PAR response should be injected.
```
const url = client.authorizationUrl(authorizationParams)
```

**Auth.js (v5.x)**

Seems this versions is using [oauth4webapi](https://github.com/panva/oauth4webapi/) and here is the sample code how PAR [works]( https://github.com/panva/oauth4webapi/blob/main/examples/par.ts)
Next auth takes care of getting the authorizationUrl [here](https://github.com/nextauthjs/next-auth/blob/main/packages/core/src/lib/actions/signin/authorization-url.ts).

**Reply by MonstraG** [reply]
Thanks!

We basically ended up dropping next-auth and just using `oauth4webapi` directly.

**Reply by akosb** [reply]
@MonstraG - this is for the record, so if anyone else is looking for this it's documented. Probably it's too late for you now, but I ended up with 'super simple' solution :)

Looking at the oauth provider of next-auth we can see that the OAuthConfig objects `authorization` attribute is an AuthorizationEndpointHandler
https://github.com/nextauthjs/next-auth/blob/v4/packages/next-auth/src/providers/oauth.ts#L118

```
export type AuthorizationEndpointHandler =
  EndpointHandler<AuthorizationParameters>
```

since EndpointHandler is an AdvancedEndpointHandler
```
export type EndpointHandler<
  P extends UrlParams,
  C = any,
  R = any,
> = AdvancedEndpointHandler<P, C, R>
```

it means that `request?: EndpointRequest<C, R, P>` is available
https://github.com/nextauthjs/next-auth/blob/v4/packages/next-auth/src/providers/oauth.ts#L54

Thus nothing stopped me to implement a `request` function for `authorization`
```
  const provider: OAuthConfig<any> = {
    ...
    authorization: {
      request: async (context) => {
      	...
      	// call my PAR endpoint and get a par url
      	...
      	return { par_responseL par_response }
      }
    }
  }
```

Values returned here (par_response) are injected into context of the request
https://github.com/nextauthjs/next-auth/blob/v4/packages/next-auth/src/providers/oauth.ts#L28

Now comes the part when make the `getAuthorizationUrl` function to use this context information
https://github.com/nextauthjs/next-auth/blob/v4/packages/next-auth/src/core/lib/oauth/authorization-url.ts#L16

```
  // if we define a request for authorization, there we get back PAR response
  if(provider.authorization?.request) {
    const res = await provider.authorization?.request({
      provider,
      params,
      checks,
      client,
    })
    const url = await client.authorizationUrl(res?.par_response)

    logger.debug("GET_AUTHORIZATION_URL", { url, cookies, provider })
    return { redirect: url, cookies }

  } else {
    const url = client.authorizationUrl(authorizationParams)

    logger.debug("GET_AUTHORIZATION_URL", { url, cookies, provider })
    return { redirect: url, cookies }

  }
```

I created a patch (https://github.com/ds300/patch-package) for this, maybe my colleagues will send it in ...

Regards, Akos
